omnigibson/envs/sb3_vec_env.py

```diff
@@ -0,0 +1,110 @@
+        for env in tmp_envs:
+            env.post_play_load()
+
+        # Now produce some functions that will make DummyVecEnv think it's creating these envs itself
+        env_fns = [lambda env_=env: env_ for env in tmp_envs]
+        super().__init__(env_fns)
+
+        # Keep track of our last reset time
+        self.last_reset_time = time.time()
+
+    def step_async(self, actions: np.ndarray) -> None:
+        # We go into this context in case the pre-step tries to call step / render
+        with og.sim.render_on_step(self.render_on_step):
+            global last_stepped_env, last_stepped_time
+
+            if last_stepped_env != self:
+                # If another environment was used after us, we need to check that we have been reset
+                # after that.
+                assert (
+                    last_stepped_time is None or self.last_reset_time > last_stepped_time
+                ), "You must call reset() before using a different environment."
+                last_stepped_env = self
+                last_stepped_time = time.time()
+
+            self.actions = actions
+            for i, action in enumerate(actions):
+                self.envs[i]._pre_step(action)
+
+    def step_wait(self) -> VecEnvStepReturn:
+        with og.sim.render_on_step(self.render_on_step):
+            # Step the entire simulation
+            og.sim.step()
+
+            for env_idx in range(self.num_envs):
+                obs, self.buf_rews[env_idx], terminated, truncated, self.buf_infos[env_idx] = self.envs[
+                    env_idx
+                ]._post_step(self.actions[env_idx])
+                # convert to SB3 VecEnv api
+                self.buf_dones[env_idx] = terminated or truncated
+                # See https://github.com/openai/gym/issues/3102
+                # Gym 0.26 introduces a breaking change
+                self.buf_infos[env_idx]["TimeLimit.truncated"] = truncated and not terminated
+
+                if self.buf_dones[env_idx]:
+                    # save final observation where user can get it, then reset
+                    self.buf_infos[env_idx]["terminal_observation"] = obs
+                    obs, self.reset_infos[env_idx] = self.envs[env_idx].reset()
+                self._save_obs(env_idx, obs)
+
+            return (
+                self._obs_from_buf(),
+                np.copy(self.buf_rews),
+                np.copy(self.buf_dones),
+                copy.deepcopy(self.buf_infos),
+            )
+
+    def reset(self):
+        with og.sim.render_on_step(self.render_on_step):
+            self.last_reset_time = time.time()
+
+            for env_idx in range(self.num_envs):
+                maybe_options = {"options": self._options[env_idx]} if self._options[env_idx] else {}
+                self.envs[env_idx].reset(get_obs=False, seed=self._seeds[env_idx], **maybe_options)
+
+            # Settle the robots
+            for _ in range(30):
+                og.sim.step()
```

Comment:
I am not sure I agree with this design. Seems fairly hacky. Why should the VectorEnvironment handle settling the robots?

Reply:
@cgokmen

Reply:
Yeah I don't love this either. The long story short is that the per-task or per-env reset code can no longer call `og.sim.step` because that will make each environment get stepped a total of n_steps * n_envs times and fully ruins speed. This is currently the only way to do this that is correct and fast (reset everything first, then settle everything). @hang-yin please update the comment to say "Settle the environments" instead. This will be fixed once we make the task classes etc. vectorized too.